### PR TITLE
Leaves - Elizabeth 

### PR DESCRIPTION
**Pull request by north108**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method #spotlight in my Work Model. It iterates through all instances of Works to find the work with the highest vote count.
Describe how you approached testing that model method. What edge cases did you come up with? | I tried to test all of the edges for the validations. I mostly worked with the validation presence so I checked that it was valid with the parameter that needed to be present, it was invalid if it wasn't there, and that it was invalid if the user left it blank.
What are `session` and `flash`? What is the difference between them? | Session is a cookie that holds on to information through redirects. Flash is a message that is displayed once a specific redirect is triggered.
What was one thing that you gained more clarity on through this assignment? | Testing
What is the Heroku URL of your deployed application? | https://en-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | The css part was definitely tough because we had to match the example, would have been more valuable to write our own bootstrap even if it was more basic looking in the end.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  post "work/:id/vote", to: "works#upvote", as: "upvote"
+
+  resources :users, only: [:show, :index] 
+
+  
+
+  get "/login", to: "users#login_form", as: "login"
+  post "/login", to: "users#login"
+  post "/logout", to: "users#logout", as: "logout"
+  get "/users/current", to: "users#current", as: "current_user"
```
Comment (dHelmgren) [inline config/routes.rb:16]: Love the restraint here!

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      </thead>
+
+      <tbody>
+        <% albums = @works.where(category: "album")%>
+        <% albums.each do |work| %> 
+            <tr>
+              <td> <%= work.votes.count %> </td>
+              <td> <%= link_to work.title, work_path(work) %> </td>
+              <td> <%= work.creator %> </td>
+              <td> <%= work.publication_year %> </td>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:24]: This code gets repeated 3 times! Could you dry this up with a view partial?

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  validates :creator, presence: true
+
+  def self.spotlight
+    works = Work.all
+
+    works.max_by do |work|
+      work.votes.count
+    end
+  end
+
```
Comment (dHelmgren) [inline app/models/work.rb:14]: It would be good to put a `sort_by_votes` in here too!

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | no flash message
Can go to "View all media" page and see three lists of works, sorted by vote | not sorted by votes!
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | no, doesn't seem to be done with seeding. Also, heads up, might want to keep an eye on your usernames.
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | Yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | no, see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | no, missing
Some controller, likely the `WorksController`, has a controller filter for finding a work | no, missing
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Good work overall! Your project is missing a few small pieces, but the majority of what you've done is spot on!
